### PR TITLE
CodeDeploy provider support for --waint-untill-deployed option

### PR DESCRIPTION
**Pull request by montaro**

### CODE REVIEW

File: lib/dpl/provider/code_deploy.rb [hunk 1]
```diff
@@ -53,17 +53,38 @@ def github_revision
...
+          log "Deployment successful."
+        else
+          error "Deployment failed."
+        end
       rescue Aws::CodeDeploy::Errors::DeploymentLimitExceededException => exception
         error(exception.message)
       end
 
+      def wait_until_deployed(deployment_id)
+        deployment = nil
```
Comment (BanzaiMan) [inline lib/dpl/provider/code_deploy.rb]: Is this necessary?

File: lib/dpl/provider/code_deploy.rb [hunk 2]
```diff
@@ -53,17 +53,38 @@ def github_revision
...
+        loop do
+          result = code_deploy.get_deployment(deployment_id: deployment_id)
+          deployment = result[:deployment_info]
```
Comment (BanzaiMan) [inline lib/dpl/provider/code_deploy.rb]: Do we know `result` would respond to `[]`?

File: lib/dpl/provider/code_deploy.rb [hunk 3]
```diff
@@ -53,17 +53,38 @@ def github_revision
...
+          break unless deployment[:status] == "Created" || deployment[:status] == "Queued" || deployment[:status] == "InProgress"
```
Comment (BanzaiMan) [inline lib/dpl/provider/code_deploy.rb]: We'll have a `NoMethodError` here if `deployment` doesn't respond to `[]` (in particular, if it's `nil`).

Comment (BanzaiMan) [inline lib/dpl/provider/code_deploy.rb]: Are these the only final state we should expect? Is there any error condition we should notice and inform the users about?

Comment (BanzaiMan) [inline lib/dpl/provider/code_deploy.rb]: If `Succeeded` is the only status we should care about, then maybe the `break` statement should be

```ruby
    break if deployment[:status] == 'Succeeded"
```

instead.

File: lib/dpl/provider/code_deploy.rb (hunk 1 above)
Comment (kfateem) [inline lib/dpl/provider/code_deploy.rb]: I believe its needed. A NameError will be thrown on line 85 if not initialized.

Comment (BanzaiMan) [inline lib/dpl/provider/code_deploy.rb]: Ah, OK. I understand now.

File: lib/dpl/provider/code_deploy.rb (hunk 2 above)
Comment (kfateem) [inline lib/dpl/provider/code_deploy.rb]: result responds to []. I've included relevant snippets from the [aws ruby sdk api](https://github.com/aws/aws-sdk-ruby/blob/master/aws-sdk-core/apis/codedeploy/2014-10-06/api-2.json) for GetDeployment:

 "GetDeployment":{
      "name":"GetDeployment",
      "http":{
        "method":"POST",
        "requestUri":"/"
      },
      "input":{"shape":"GetDeploymentInput"},
      "output":{"shape":"GetDeploymentOutput"},
      "errors":[
        {
          "shape":"DeploymentIdRequiredException",
          "exception":true
        },
        {
          "shape":"InvalidDeploymentIdException",
          "exception":true
        },
        {
          "shape":"DeploymentDoesNotExistException",
          "exception":true
        }
      ]
    }

"GetDeploymentOutput":{
      "type":"structure",
      "members":{
        "deploymentInfo":{"shape":"DeploymentInfo"}
      }
    }

File: lib/dpl/provider/code_deploy.rb (hunk 1 above)
Comment (BanzaiMan) [inline lib/dpl/provider/code_deploy.rb]: You'd probably want

```ruby
deployment = {}
```

which agrees with the general type that `loop` is expected to assign to `deployment`. (I understand that the value is overwritten, but it reads better to me.)

File: lib/dpl/provider/code_deploy.rb (hunk 3 above)
Comment (kfateem) [inline lib/dpl/provider/code_deploy.rb]: deployment should always respond to [ ].

result[:deployment_info] will always return a structure.

Comment (kfateem) [inline lib/dpl/provider/code_deploy.rb]: I believe the error states are the 3 exceptions on GetDeployment:
```json
"errors":[
        {
          "shape":"DeploymentIdRequiredException",
          "exception":true
        },
        {
          "shape":"InvalidDeploymentIdException",
          "exception":true
        },
        {
          "shape":"DeploymentDoesNotExistException",
          "exception":true
        }
```

Currently the errors are re-thrown.

Comment (kfateem) [inline lib/dpl/provider/code_deploy.rb]: The possible status values are:

```json
"DeploymentStatus":{
      "type":"string",
      "enum":[
        "Created",
        "Queued",
        "InProgress",
        "Succeeded",
        "Failed",
        "Stopped"
      ]
    }
```
The deployment should wait if status is "Created",  "Queued", or "InProgress".
Lines 67-71 will test for success:
```ruby
  if deployment[:status] == 'Succeeded'
+          log "Deployment successful."
+        else
+          error "Deployment failed."
+        end
```
Is this sufficient?

File: lib/dpl/provider/code_deploy.rb (hunk 1 above)
Comment (kfateem) [inline lib/dpl/provider/code_deploy.rb]: Makes sense. Will update.

File: lib/dpl/provider/code_deploy.rb (hunk 3 above)
Comment (BanzaiMan) [inline lib/dpl/provider/code_deploy.rb]: Ah, I see.

Logically, though, it seems more natural to me to "break out of the `loop` when the deployment reached its final state (`Succeeded`, `Failed`, or `Stopped`)". The current logic reads "do not break if the state is one of these", which requires an extra step in my mind.

The handling of the deployment status in `push_app` seems OK.

Comment (kfateem) [inline lib/dpl/provider/code_deploy.rb]: Good point. Adjust loop to break on final state for better readability.

**Comment by BanzaiMan** [comment]
Once https://travis-ci.org/travis-ci/dpl/builds/65953056 passes, you can test it with:

```yaml
deploy:
  provider: code_deploy
  ⋮
  edge: true
```

**Pull request merged**